core/src/main/scala/scalaz/Foldable.scala

```diff
@@ -364,15 +365,13 @@ trait Foldable[F[_]]  { self =>
    * Selects groups of elements that satisfy p and discards others.
    */
   def selectSplit[A](fa: F[A])(p: A => Boolean): List[NonEmptyList[A]] =
-    foldRight(fa, (List[NonEmptyList[A]](), false))((a, xb) => xb match {
-      case (x, b) => {
-        val pa = p(a)
-        (if (pa)
-          if (b)
-            (a <:: x.head) :: x.tail else
-            NonEmptyList(a) :: x
-        else x, pa)
-      }
+    foldRight(fa, (List.empty[NonEmptyList[A]], true))((a, lb) => (lb, p(a)) match {
```

Comment:
The cases below where `p(a)` is `true` and where it is `false` are disjoint. Checking `if (p(a))` first would save us allocating the `(lb, p(a))` tuple.

Reply:
Right, will do.